### DISCUSSION TITLE
ref get and set accessors, ref field(per c# 11 semi auto property field keyword)

### DISCUSSION THREAD

**Discussion by emceelovin** [dotnet/csharplang#5598] [Language Ideas]
In reference to [issue](https://github.com/dotnet/runtime/issues/63111)

To keep everything short, as most of what I have to say is in that issue reference, property get and set accessors should allow by reference syntax. Since this is merely a syntax extension, it shouldn't break anything. It mainly allows a significant improvement in speed due to resulting code generation. Currently, you can return by reference via the get accessor only, but the syntax breaks the logic IMO.

Current syntax:
public ref LargeValueType LargeValueTypeProperty { get; set; }

Proposed syntax:
public LargeValueType LargeValueTypeProperty { ref get; ref set; }

public LargeValueType LargeValueTypeProperty
{
    ref get => ref _manualBackingField; // allow reference to manual backing field to avoid useless copying

    get => FunctionCall(ref field); // allow reference to auto generated backing field to avoid useless copying

    ref set => field = value; // value is automatically by ref due to ref set syntax. allows direct heap to heap copying - avoiding the stack - in the cases where  the ValueType is boxed or nested in a reference type.

    ref get;

    ref set;

    // implies
    ref LargeValueType get_LargeValueTypeProperty() => ref LargeValueTypePropertyk<>_BackingField;
    
    void set_LargeValueTypeProperty(ref LargeValueType value) => LargeValueTypePropertyk<>_BackingField = value;
}

old code:
OldByValueFunctionParameter(object.PropertyWithUpdatedRefGetAccessor); // would just cause a copy to be pushed onto the stack.
NewAndImprovedFunction(ref object.PropertyWithUpdatedRefGetAccessor); // would be able to take advantage of get by ref

With the current syntax, the implication is that a ref LargeValueType is going to be returned. E.g. a manual/auto generated 'private ref LargeValueType LargeValueTypePropertyk<>__BackingField is what I am going to receive via the get accessor. Not a reference to the actual backing field itself. Due to the nature/design of properties. If that makes sense. Which I don't even think is legal except for in a ref struct? Either way, I can't think of any valid reason to not allow this. Mostly due to the fact that it just logically makes sense, and no matter how a property is generated - auto or manual, it's backed by a field. Which completely allows access by address all the way down to the IL/JIT level. I am more than happy to take any criticism. By all means. I just feel this would mainly improve performance, but it starts at a language change.

**Comment by bernd5** [comment]
What you referred to as "Current syntax" is actually illegal (auto properties can't be by-ref)

> Current syntax:
> ```csharp
> public ref LargeValueType LargeValueTypeProperty { get; set; }
> ```

But what you can write today is:
```csharp
private static LargeValueType largeValueTypeField
public ref LargeValueType LargeValueTypeProperty { get => ref largeValueTypeField; }
```

To support by-ref-auto-properties we would need first by-ref-fields - which are currently not allowed.
In the dotnet runtime we have a special type [`ByReference<T>`](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/ByReference.cs) which makes it possible to implement e.g. [System.Span](https://github.com/dotnet/runtime/blob/e571c3b1348febce60153004760504b9c330f98c/src/libraries/System.Private.CoreLib/src/System/Span.cs#L26).
The type `ByReference<T>` is `internal` and uses some runtime intrinsic.

**Reply by emceelovin** [reply]
> What you referred to as "Current syntax" is actually illegal
> 
> Current syntax:
> 
> ```cs
> public ref LargeValueType LargeValueTypeProperty { get; set; }
> ```
> 
> But what you can write today is:
> 
> ```cs
> private static LargeValueType largeValueTypeField
> public ref LargeValueType LargeValueTypeProperty { get => ref largeValueTypeField; }
> ```
> 
> To support by-ref-auto-properties we would need first by-ref-fields - which are currently not allowed. In the dotnet runtime we have a special type `ByReference<T>` which uses some runtime intrinsic, see e.g. [System.Span](https://github.com/dotnet/runtime/blob/e571c3b1348febce60153004760504b9c330f98c/src/libraries/System.Private.CoreLib/src/System/Span.cs#L26).

Thanks for the reply! You are indeed correct. I quickly tossed the code together in my original post. I apologize. what i wrote is indeed illegal due to the set accessor. i forgot to leave that out when you have a ref property via current syntax. which is part of the reason for the post. you cannot have a by ref set accessor with the current language specification. Bear with me also I'm replying via mobile.

**Comment by emceelovin** [comment]
you wouldn't need by ref fields to support by ref accessing. auto generated properties generate a backing field. access to them by reference is completely possible via the current runtime. i'm familiar with how Span was able to be implemented, but the same logic is not required to implement this. field access by address is already allowed in the runtime via opcode ldflda. Span logic doesn't make any sense here at all. i do not want a by ref field as I've said before. like you and i both mention, it's not allowed currently. this whole idea is relative to the accessing of actual property data itself. not the type of property.

**Reply by bernd5** [reply]
Ah, you want to add an additional by ref accessor - not just `get` and `set` \ `init`?

So you could write:

```csharp
class Foo{
   private int data;
   public int Data
   {
       get => data;
       ref get => ref data;
       set => data = value;
   }
}
```

Why not simply:
```csharp
class Foo{
   private int data;
   public int Data
   {
       get => data;
       set => data = value;
   }
   public ref int DataRef => ref data;
}
```

**Reply by emceelovin** [reply]
> Ah, you want to add an additional by ref accessor - not just `get` and `set` \ `init`?
> 
> So you could write:
> 
> ```cs
> class Foo{
>    private int data;
>    public int Data
>    {
>        get => data;
>        ref get => ref data;
>        set => data = value;
>    }
> }
> ```
> 
> Why not simply:
> 
> ```cs
> class Foo{
>    private int data;
>    public int Data
>    {
>        get => data;
>        set => data = value;
>    }
>    public ref int DataRef => ref data;
> }
> ```

Exactly, but a less syntactical better shortcut approach. I was saying how I think the syntax should be applied to the get/set syntax, and not the actual property itself. As that implies a ByReference<T> field(to me). I believe it makes more sense, and will not break anything, if its an extension of get/set syntax. Seeing as how that's where access is defined. Also, the set accessor is more so the problem. You can have a by ref get accessor(with the syntax you have mentioned) on an auto generated property, but not both a by ref get and set accessor on an auto generated property. Depending on how hot this codepath is, a server via serialization or Direct3D three dimensional vector, accessing a decently sized value type constantly adds up. Assuming both value types are on the heap in some way - nested in reference types, boxed, async functions, etc, heap -> cpu -> stack -> cpu -> heap is the abridged process that has to take place to move the data. With a ref set accessor, you can double the performance by eliminating the stacks involvement. Let me illustrate further with some current csharp 10 code(i'm pretty new to github, but i figured out how to format replies better now):

```cs
class Foo
{
    /*
     * achieves the goal, but with way too much syntax, and does not allow set accessor processing of source data.
     * has to be done after you retrieve the property by reference; inconvenient.
     */
    public ref LargeValueType ManualGeneratedByRefProperty { get => ref _largeValueType; }

    // goal
    public LargeValueType AutoGeneratedByRefAccessorProperty { ref get; ref set;}


    // by ref accessing of backing field due to manual generation. somewhat halfway i suppose
    public LargeValueType LargeValueTypeProp
    {
        get => _largeValueType;

        set => ByRefCallee(ref _largeValueType, value);
    }

    // causes an unnecessary copy
    static public void ByRefCallee(ref LargeValueType destination, LargeValueType source)
        =>
        destination = source;

    // extended accessor syntax access to backing field via field keyword in csharp 11 proposal
    public LargeValueType LargeValueTypeRef
    {
        /* (auto generated get accessor high-level pseudo)
         * ref var autoRef = ref object.LargeValueTypeRef; with this(and the csharp 10 version), anytime the property is accessed,
         * you're accessing the exact location of that field in memory. no more, no less.
         */
        ref get => ref field;

        /*
         * ref should be allowed on the field keyword the same way you're able to access a manually generated backing field.
         * we're given access to the backing field, why not allow by reference access to avoid additional copies etc? it's
         * the same as if i had created a backing field myself. field == _largeValueType.
         * 
         * also allows object.LargeValueTypeRef = ref sourceObject.LargeValueTypeRef; direct memory to memory. no stack involved
         */
        ref set
        {
            // any necessary computation someone wants to do
            ByRefCalleeRefSource(ref field, ref value);
        }

        /*
         *  (csharp 11 auto generated set accessor high-level pseudo)
         */
        ref set => field = value;

        /*
         * internal generated
         */
        ref LargeValueType get_LargeValueTypeRef() => ref field;

        void set_LargeValueTypeRef(ref LargeValueType value) => field = value;
    }

    static public void ByRefCalleeRefSource(ref LargeValueType destination, ref LargeValueType source)
        =>
        destination = source;

    LargeValueType _largeValueType;
}
```
